File: learning-core/executable/idempotent-message/go/idempotent_message.go
package idempotentmessage

import "errors"

type State struct {
	Processed map[string]bool
	Total     int
}

func NewState() *State {
	return &State{Processed: map[string]bool{}}
}

func Handle(state *State, messageID string, amount int) (bool, error) {
	if messageID == "" {
		return false, errors.New("message id is required")
	}
	if state.Processed[messageID] {
		return false, nil
	}
	state.Total += amount
	state.Processed[messageID] = true
	return true, nil
}

File: learning-core/executable/idempotent-message/cpp/idempotent_message.cpp
#include <stdexcept>
#include <string>
#include <unordered_set>

struct ConsumerState {
    std::unordered_set<std::string> processed_ids;
    int total{0};
};

bool handle(ConsumerState& state, const std::string& message_id, int amount) {
    if (message_id.empty()) {
        throw std::invalid_argument("message id is required");
    }
    if (state.processed_ids.contains(message_id)) {
        return false;
    }
    state.total += amount;
    state.processed_ids.insert(message_id);
    return true;
}
